File: C++ Value-Iteration base Implement/DeterministicGridWorld.cpp
#include "DeterministicGridWorld.hpp"
#include "algorithm"


// Constractor overloading
//----------------------------
GridWorld::GridWorld( int row, int col, std::pair<int,int> start_state ){

    this->rows = row;
    this->cols = col;
    this->current_state = start_state;

    // initializing States >>>
    for( int r=0; r<row ; r++ ){
        for( int c=0; c<col ; c++ ){
            this->states.push_back( {r,c} );
            this->actions.push_back({});
            this->rewards.push_back(0.0);
        }
    }
}

GridWorld::GridWorld( int row, int col, float cost, std::pair<int,int> start_state ){

    this->rows = row;
    this->cols = col;
    this->current_state = start_state;

    // initializing States >>>
    for( int r=0; r<row ; r++ ){
        for( int c=0; c<col ; c++ ){
            this->states.push_back( {r,c} );
            this->actions.push_back({});
            this->rewards.push_back(cost);
        }
    }
}


// Regular Getters
//---------------------
int GridWorld::get_colsize(void){ return this->cols;}
int GridWorld::get_rowsize(void){ return this->rows;}
std::pair<int,int> GridWorld::get_currentState(void){ return this->current_state;}

std::vector< float > GridWorld::get_AllRewards(void){ return this->rewards;}
std::vector< std::pair<int,int> > GridWorld::get_AllStates(void){ return this->states;}
std::vector< std::vector<std::string> > GridWorld::get_AllActions(void){ return this->actions;}


// Current State Actions getter.
//--------------------------------
std::vector<std::string> GridWorld::get_StateActions(void){
    
    int state_index = this->current_state.first * this->cols + this->current_state.second ;
    return this->actions[state_index];
}

// Environment Transition function: T(s,a)-> r,s'
// In model Base RL learning, transition func will not be used.
//------------------------------------------------
float GridWorld::transition( std::string action ){

    int i = this->current_state.first;
    int j = this->current_state.second;
    int index = i * this->cols + j ;

    std::vector<std::string> avl_actions = this->actions[index];

    if( std::find( avl_actions.begin(), avl_actions.end(), action ) == avl_actions.end() ){
        throw GridWorld::InvalidActionException();
    }

    if ( action == "U" )
        i-- ;
    else if ( action == "D" )
        i++ ;
    else if ( action == "R" )
        j++ ;
    else if ( action == "L" )
        j-- ;
    
    this->current_state = {i,j};
    return this->rewards[ i*this->cols+j ];
}

// Environment Transition Probability function: p(s',r|s,a)-> 0/1
//-----------------------------------------------------------------
float GridWorld::probability( std::pair<int,int> to_state, std::pair<int,int> from_state, std::string action ){

    int i = 0;
    int j = 0;

    if      ( action == "U" ) i-- ;
    else if ( action == "D" ) i++ ;
    else if ( action == "R" ) j++ ;
    else if ( action == "L" ) j-- ;
    else 
        throw GridWorld::InvalidActionException();
    
    int index = from_state.first * this->cols + from_state.second ;
    std::vector< std::string > avl_actions = this->actions[index];
    if( std::find( avl_actions.begin(), avl_actions.end(), action ) == avl_actions.end() ){
        throw GridWorld::InvalidActionException();
    }

    if( (from_state.first + i == to_state.first) && (from_state.second + j == to_state.second) ){
        return 1.0;
    }
    return 0.0;
}


// Some God Mode methods prototype
//----------------------------------
bool GridWorld::is_terminal( std::pair<int,int> state){
    std::vector<std::string> avl_actions = this->actions[ state.first * this->cols + state.second ];
    return avl_actions.empty();
}


// Game-over check-method.
//--------------------------
bool GridWorld::game_over(void){
    std::vector<std::string> avl_actions = this->actions[ this->current_state.first * this->cols + this->current_state.second ];
    return avl_actions.empty();
}


// Updating grid-world's actions & rewards config.
//---------------------------------------------------
void GridWorld::set_config( std::vector< std::vector<std::string> > actions, std::vector< float > rewards){
    this->actions = actions;
    this->rewards = rewards;
}



//  Standard GW creation.
// -------------------------------
GridWorld standard_GW(void){
    int row = 3;
    int col = 4;

    std::pair<int,int> start_state = {2,0};
    GridWorld gw = GridWorld(row, col, start_state);

    return gw;
}

GridWorld standard_GW(float cost){
    int row = 3;
    int col = 4;

    std::pair<int,int> start_state = {2,0};
    GridWorld gw = GridWorld(row, col, cost, start_state);

    return gw;
}

File: Python Policy-Evaluation base Implement/Model Base Deterministic/Deterministic_Grid_World.py
import pickle
import numpy as np

# Deterministic Grid-world class:
#---------------------------------------------------------------

# Deterministic transition function -> if you choose U, just move up.
# Deterministic rewards -> you will get same reward from any way you reach s' form s.
class GridWorld :
    
    def __init__(self, rows, cols, start_state):
        
        self.cols:int = cols
        self.rows:int = rows
        
        # Agent Current State = (i,j)
        self.current_state: tuple = start_state
        
        # States = [ (i,j) ]
        self.states : list = []
        
        # Actions = { (i,j):["U","D","R","L"] }
        self.actions: dict = {}
        
        # Rewards = { (i,j):float }
        self.rewards: dict = {}
        
        self.__initialize_states()
        
    
    
    # Current State Actions getter.
    #----------------------------------
    def get_actions(self):
        return self.actions.get(self.current_state)
    
    
    # Environment Transition function: T(s,a)-> r,s'
    #-----------------------------------------------
    def transition(self, action:str ) -> float :
        
        i,j = self.current_state
        
        if action in self.actions.get(self.current_state):
            if action == "U":
                if i > 0 :
                    i -= 1
            elif action == "D":
                if i < self.rows-1 :
                    i += 1
            elif action == "L":
                if j > 0 :
                    j -= 1
            elif action == "R":
                if j < self.cols-1:
                    j += 1
        
        self.current_state = (i,j)
        return self.rewards.get(self.current_state,0)
    
    
    # Environment Transition Probability function: p(s',r|s,a)-> 0/1
    #-----------------------------------------------------------------
    def probability(self, to_state, from_state, action ):
        
        i,j = 0,0
        if   action == "U": i -= 1
        elif action == "D": i += 1
        elif action == "R": j += 1
        elif action == "L": j -= 1
        else : return 0.0
        
        if action in self.actions.get(from_state,[]):
            if tuple((from_state[0]+i, from_state[1]+j)) == to_state :
                return 1.0
            else :
                return 0.0
        else : 
            return 0.0
    
    
    # Some God Mode methods prototype
    #---------------------------------
    def is_terminal(self, state ):
        if self.actions.get(state,[]) == [] :
            return True
        else:
            return False
    
    def undo_action(self, action): pass
    def get_next_state(self,state,action): pass
    def set_state(self, new_state): pass
    #=================================
    
    # Game-over check-method.
    #----------------------------
    def game_over(self):
        if self.actions.get(self.current_state,[]) == [] :
            return True
        else:
            return False
    
    
    # Updating grid-world's actions & rewards config.
    #--------------------------------------------
    def set_config( self, actions:dict, rewards: dict ):
        self.actions = actions
        self.rewards = rewards
    
    
    # Initialize Grid-world States.
    #----------------------------------
    def __initialize_states(self):
        
        for r in range(self.rows):
            for c in range(self.cols):
                self.states.append((r,c))
        
    
#======================================================================================================
#======================================================================================================

# Global functions:
#============================

# saving GW configs into file.
#-----------------------------
def save_env(Gw, filename):
    with open(filename, 'wb') as f:
        pickle.dump(Gw, f)


# Loading GW configs from file.
#-------------------------------
def load_env(filename):
    with open(filename, 'rb') as f:
        return pickle.load(f)


# Standard GW creation.
#-------------------------------
def standard_GW():
    r = 3
    c = 4
    start_state = (2,0)
    g = GridWorld(r,c,start_state)
    
    rewards = {}
    for i in range(r):
        for j in range(c):
            rewards[(i,j)] = -0.1
    rewards[(0,3)] = 1.0
    rewards[(1,3)] = -1
    rewards[(1,1)] = 0
    
    
    actions = {
        (0,0):["D","R"],
        (0,1):["R","L"],
        (0,2):["R","L","D"],
        (1,0):["U","D"],
        (1,2):["R","U","D"],
        (2,0):["U","R"],
        (2,1):["R","L"],
        (2,2):["U","R","L"],
        (2,3):["L","U"]
    }
    g.set_config(actions,rewards)
    return g

